### PR TITLE
Serialization

### PR DESCRIPTION
**Pull request by iskandr**
We previously had implementations of `__getstate__` and `__setstate__` which (1) didn't work for non-human genomes and (2) used fragile positional arguments. This PR:

* Creates a `Serializable` base class for `Genome`, `EnsemblRelease`, `Gene`, `Transcript`, and `Exon`, `Species`. This allows us to share serialization/flattening code between these objects with copy/paste madness. 
* Adds `to_json`/`from_json` and `to_dict`/`from_dict` methods to all serializable objects. 
* Moves `normalize_chromosome` and `normalize_strand` to a separate module (no changes to actual code)
* Moved global state from `species.py` onto the `Species` class

---
This change is [Reviewable](https://reviewable.io/reviews/hammerlab/pyensembl/154)

### CODE REVIEW

File: pyensembl/exon.py [hunk 1]
```diff
@@ -35,12 +35,6 @@ def __str__(self):
         return "Exon(exon_id=%s, gene_name=%s, contig=%s, start=%d, end=%s)" % (
             self.id, self.gene_name, self.contig, self.start, self.end)
 
-    def __repr__(self):
```
Comment (tavinathanson) [inline pyensembl/exon.py]: What happened to `repr`?

File: pyensembl/exon.py [hunk 2]
```diff
@@ -49,3 +43,13 @@ def __eq__(self, other):
...
             self.id == other.id)
+
+    def __hash__(self):
+        return hash(self.id)
+
+    def to_dict(self):
+        state_dict = Locus.to_dict(self)
+        state_dict["exon_id"] = self.id
+        state_dict["gene_name"] = self.gene_name
+        state_dict["gene_id"] = self.gene_id
```
Comment (tavinathanson) [inline pyensembl/exon.py]: `strand`?

Comment (tavinathanson) [inline pyensembl/exon.py]: Nevermind, covered in `Locus`

File: pyensembl/exon.py (hunk 1 above)
Comment (iskandr) [inline pyensembl/exon.py]: All of our implementations just call `str` so I made that the default in `Serializable`. Maybe that object needs to be renamed to something like `SaneDefaultsForClassesWhichImplementToDict`.

File: pyensembl/ensembl_release.py [hunk 3]
```diff
@@ -89,10 +86,14 @@ def __eq__(self, other):
...
-        self.__init__(*fields)
+    def to_dict(self):
+        return {
+            "release": self.release,
+            "species": self.species.to_dict(),
+            "server": self.server
+        }
 
+    @classmethod
+    def _initialize_nested_objects_in_state_dict(cls, state_dict):
```
Comment (tavinathanson) [inline pyensembl/ensembl_release.py]: Agree that this feels very automate-able but fine for now.

File: pyensembl/gene.py [hunk 4]
```diff
@@ -34,24 +32,15 @@ def __init__(
             biotype,
             genome,
             require_valid_biotype=True):
-        """
-        genome is a Genome object.
-        """
+        LocusWithGenome.__init__(self, contig, start, end, strand, genome)
         self.id = gene_id
-        self.genome = genome
-        self.db = genome.db
```
Comment (tavinathanson) [inline pyensembl/gene.py]: `db` wasn't used?

File: pyensembl/genome.py [hunk 5]
```diff
@@ -117,23 +118,21 @@ def __init__(
...
-            self.cache_directory_path
-        ]
-
-    def __setstate__(self, fields):
-        self.__init__(*fields)
+    def to_dict(self):
+        """
+        Returns a dictionary of the essential fields of this Genome.
+        """
+        return dict(
```
Comment (tavinathanson) [inline pyensembl/genome.py]: This is valid Python?

Comment (tavinathanson) [inline pyensembl/genome.py]: Whoa, apparently it is, crazy...

File: pyensembl/gene.py (hunk 4 above)
Comment (tavinathanson) [inline pyensembl/gene.py]: Oh, I see, `db` is in `Genome`

File: pyensembl/locus_with_genome.py [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+        return dict(
+            contig=self.contig,
+            start=self.start,
+            end=self.end,
+            strand=self.strand,
+            genome=self.genome._object_to_primitive_types())
+
+    @classmethod
+    def _initialize_nested_objects_in_state_dict(cls, state_dict):
+        state_dict["genome"] = cls._object_from_primitive_types(state_dict["genome"])
```
Comment (tavinathanson) [inline pyensembl/locus_with_genome.py]: Add a brief comment explaining what's happening here for when we look at it in a few weeks and forget everything?

File: pyensembl/normalization.py [hunk 7]
```diff
@@ -0,0 +1,65 @@
+# Copyright (c) 2016. Mount Sinai School of Medicine
```
Comment (tavinathanson) [inline pyensembl/normalization.py]: https://github.com/tavinathanson/banana saves the day by informing me that this code is not new!

File: pyensembl/serializable.py [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+    """
+    Base class for all PyEnsembl objects which provides default
+    methods such as to_json, from_json, __reduce__, and from_dict
+    and relies only on (1) a user-defined to_dict method and (2) that the
+    keys of to_dict match the arguments to __init__.
+    """
+
+    def __str__(self):
+        fields_str = ", ".join(
+            "%s=%s" % (k, v) for (k, v) in sorted(self.to_dict().items()))
```
Comment (tavinathanson) [inline pyensembl/serializable.py]: I could see a use-case for not wanting to rely on sorting, but rather on some instrinsic ordering (e.g. ID first, name second, etc.). Any interest in using `OrderedDict`s instead?

File: pyensembl/serializable.py [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+            return False
+        self_dict = self.to_dict()
+        other_dict = other.to_dict()
+        if len(self_dict) != other_dict(other_dict):
+            return False
+        self_keys = set(self_dict.keys())
+        other_keys = set(other_dict.keys())
+        if self_keys != other_keys:
+            return False
+        for key in self_keys:
```
Comment (tavinathanson) [inline pyensembl/serializable.py]: Curious why not `self_dict == other_dict` here?

File: pyensembl/serializable.py [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+            if self_dict[key] != other_dict[key]:
+                return False
+        return True
+
+    def to_dict(self):
+        raise NotImplementedError("Method to_dict() not implemented for %s" % (
```
Comment (tavinathanson) [inline pyensembl/serializable.py]: Add warning about needing the keys to match the `__init__` args?

Comment (tavinathanson) [inline pyensembl/serializable.py]: Also, remind me what happens if they don't?

File: pyensembl/serializable.py [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+            - name of the class
+
+        The class can be reconstructed from these two strings by calling
+        _tuple_to_class.
+        """
+        return cls.__module__, cls.__name__
+
+    def _object_to_primitive_types(self):
+        """
+        Given an instance of a Python object, returns a tuple whose
```
Comment (tavinathanson) [inline pyensembl/serializable.py]: I'm getting confused between the `tuple` here and the one above. Maybe rename `_class_to_tuple` to `_class_to_full_name_tuple` and `_tuple_to_class` to `_full_name_tuple_to_class`?

File: pyensembl/serializable.py [hunk 12]
```diff
@@ -0,0 +1,143 @@
...
+        first element is a primitive representation of the class and whose
+        second element is a dictionary of instance data.
+        """
+        class_representation = self._class_to_tuple()
+        state_dict = self.to_dict()
+        return (class_representation, state_dict)
+
+    @classmethod
+    def _object_from_primitive_types(cls, class_and_data_pair):
```
Comment (tavinathanson) [inline pyensembl/serializable.py]: Nit: maybe `_primitive_types_to_object` to be consistent with e.g. `_class_to_tuple` vs. `_tuple_to_class`?

File: pyensembl/serializable.py [hunk 13]
```diff
@@ -0,0 +1,143 @@
...
+        """
+        class_representation, state_dict = class_and_data_pair
+        subclass = cls._tuple_to_class(class_representation)
+        return subclass.from_dict(state_dict)
+
+    @classmethod
+    def _initialize_nested_objects_in_state_dict(cls, state_dict):
+        """
+        Nested serializable objects will be represented as dictionaries so we
+        allow manual reconstruction of those objects in this method.
```
Comment (tavinathanson) [inline pyensembl/serializable.py]: Clarify that it does nothing unless overrided?

File: pyensembl/serializable.py (hunk 8 above)
Comment (iskandr) [inline pyensembl/serializable.py]: You mean just use the order of `items()` but have `to_dict()` return an `OrderedDict`?

File: pyensembl/serializable.py (hunk 9 above)
Comment (iskandr) [inline pyensembl/serializable.py]: Ha, that works too.

File: pyensembl/species.py [hunk 14]
```diff
@@ -55,51 +89,33 @@ def __str__(self):
...
+
+    @classmethod
+    def from_dict(cls, state_dict):
+        return cls._latin_names_to_species[state_dict["latin_name"]]
+
+    def __reduce__(self):
+        """
+        Normally unpickling Species would create two distinct objects for
+        the same latin name, so we have to specify a custom method
+        for pickling.
```
Comment (tavinathanson) [inline pyensembl/species.py]: Clarify that this code will result in `get`ing the old `Species`?

File: test/data.py [hunk 15]
```diff
@@ -90,3 +90,43 @@ def data_path(name):
...
+    gtf_path_or_url=MOUSE_ENSMUSG00000017167_PATH,
+    transcript_fasta_path_or_url=MOUSE_ENSMUSG00000017167_TRANSCRIPT_FASTA_PATH,
+    protein_fasta_path_or_url=MOUSE_ENSMUSG00000017167_PROTEIN_FASTA_PATH)
+
+def setup_init_custom_mouse_genome():
+    """
+    If a unit test needs to start from a cleared cache, add this to the test
+    setup.
+    """
+    global custom_mouse_genome_grcm38_subset
```
Comment (tavinathanson) [inline test/data.py]: Get rid of `global`?

**Comment by tavinathanson** [comment]
LGTM % comments 👍

File: pyensembl/locus_with_genome.py (hunk 6 above)
Comment (iskandr) [inline pyensembl/locus_with_genome.py]: 👍

File: pyensembl/serializable.py (hunk 10 above)
Comment (iskandr) [inline pyensembl/serializable.py]: You'll get an error when you try to reconstruct the object (since unknown arg names will get passed to __init__).

File: test/data.py (hunk 15 above)
Comment (iskandr) [inline test/data.py]: 👍
